File: components/brave_wallet_ui/utils/network-utils.ts
// Copyright (c) 2022 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// you can obtain one at https://mozilla.org/MPL/2.0/.

import { assertNotReached } from 'chrome://resources/js/assert.js'

// types
import { BraveWallet, SerializableSolanaTxData } from '../constants/types'

export const emptyNetwork: BraveWallet.NetworkInfo = {
  chainId: '',
  chainName: '',
  activeRpcEndpointIndex: 0,
  rpcEndpoints: [],
  blockExplorerUrls: [],
  iconUrls: [],
  symbol: '',
  symbolName: '',
  decimals: 0,
  coin: BraveWallet.CoinType.ETH,
  supportedKeyrings: []
}

export const getNetworkInfo = (
  chainId: string,
  coin: BraveWallet.CoinType,
  list: BraveWallet.NetworkInfo[]
) => {
  for (let it of list) {
    if (it.chainId === chainId && it.coin === coin) {
      return it
    }
  }
  return emptyNetwork
}

export const networkSupportsAccount = (
  network: Pick<BraveWallet.NetworkInfo, 'coin' | 'supportedKeyrings'>,
  accountId: BraveWallet.AccountId
) => {
  return (
    network.coin === accountId.coin &&
    network.supportedKeyrings.includes(accountId.keyringId)
  )
}

export const filterNetworksForAccount = (
  networks: BraveWallet.NetworkInfo[],
  accountId: BraveWallet.AccountId
): BraveWallet.NetworkInfo[] => {
  if (!networks) {
    return []
  }
  return networks.filter((network) =>
    networkSupportsAccount(network, accountId)
  )
}

export const getTokensNetwork = (
  networks: BraveWallet.NetworkInfo[],
  token: BraveWallet.BlockchainToken
): BraveWallet.NetworkInfo => {
  if (!networks) {
    return emptyNetwork
  }

  const network = networks.filter((n) => n.chainId === token.chainId)
  if (network.length > 1) {
    return (
      network?.find(
        (n) => n.symbol.toLowerCase() === token.symbol.toLowerCase()
      ) ?? emptyNetwork
    )
  }

  return network[0] ?? emptyNetwork
}

export type TxDataPresence = {
  ethTxData?: Partial<BraveWallet.TxDataUnion['ethTxData']> | undefined
  ethTxData1559?: Partial<BraveWallet.TxDataUnion['ethTxData1559']> | undefined
  solanaTxData?:
    | Partial<BraveWallet.TxDataUnion['solanaTxData']>
    | SerializableSolanaTxData
    | undefined
  filTxData?: Partial<BraveWallet.TxDataUnion['filTxData']> | undefined
  btcTxData?: Partial<BraveWallet.TxDataUnion['btcTxData']> | undefined
  zecTxData?: Partial<BraveWallet.TxDataUnion['zecTxData']> | undefined
}

export const getCoinFromTxDataUnion = <T extends TxDataPresence>(
  txDataUnion: T
): BraveWallet.CoinType => {
  if (txDataUnion.ethTxData || txDataUnion.ethTxData1559) {
    return BraveWallet.CoinType.ETH
  }
  if (txDataUnion.filTxData) {
    return BraveWallet.CoinType.FIL
  }
  if (txDataUnion.solanaTxData) {
    return BraveWallet.CoinType.SOL
  }
  if (txDataUnion.btcTxData) {
    return BraveWallet.CoinType.BTC
  }
  if (txDataUnion.zecTxData) {
    return BraveWallet.CoinType.ZEC
  }

  assertNotReached('Unknown transaction coin')
}

export const reduceNetworkDisplayName = (name?: string) => {
  if (!name) {
    return ''
  }
  return name.split(' ')[0]
}

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/NetworkUtils.java
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet.util;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.chromium.brave_wallet.mojom.BraveWalletConstants;
import org.chromium.brave_wallet.mojom.CoinType;
import org.chromium.brave_wallet.mojom.NetworkInfo;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class NetworkUtils {

    public static Comparator<NetworkInfo> sSortNetworkByPriority =
            Comparator.comparingInt(c -> getCoinRank(c.coin));

    // Solana first
    private static int getCoinRank(@CoinType.EnumType int coin) {
        switch (coin) {
            case CoinType.SOL:
                return 0;
            case CoinType.ETH:
                return 1;
            case CoinType.FIL:
                return 2;
            case CoinType.BTC:
                return 3;
            default:
                return Integer.MAX_VALUE;
        }
    }

    public static class Filters {
        public static boolean isSameNetwork(NetworkInfo network, String chainId, int coin) {
            return network.chainId.equals(chainId) && network.coin == coin;
        }

        public static boolean isSameNetwork(NetworkInfo network1, NetworkInfo network2) {
            return isSameNetwork(network1, network2.chainId, network2.coin);
        }

        public static boolean isLocalNetwork(NetworkInfo network) {
            return network.chainId.equals(BraveWalletConstants.LOCALHOST_CHAIN_ID);
        }
    }

    /**
     * Returns {@code true} if two networks are equal.The comparison is done checking chain name,
     * chain ID, coin, symbol and symbol name.
     */
    public static boolean areEqual(
            @NonNull final NetworkInfo network1, @NonNull final NetworkInfo network2) {
        return network1.chainName.equals(network2.chainName)
                && network1.chainId.equals(network2.chainId)
                && network1.coin == network2.coin
                && network1.symbol.equals(network2.symbol)
                && network1.symbolName.equals(network2.symbolName);
    }

    public static boolean isAllNetwork(@Nullable final NetworkInfo networkInfo) {
        if (networkInfo == null) return false;
        return networkInfo.chainId.equals("all");
    }

    @SuppressWarnings("NoStreams")
    public static List<NetworkInfo> nonTestNetwork(List<NetworkInfo> networkInfos) {
        if (networkInfos == null) return Collections.emptyList();
        return networkInfos.stream()
                .filter(network -> !WalletConstants.KNOWN_TEST_CHAIN_IDS.contains(network.chainId))
                .collect(Collectors.toList());
    }

    /**
     * Gets the network info object of given chainId and symbol.
     *
     * @param networks All networks available.
     * @param chainId Chain Id of the network to be found.
     * @param coin Coin type of the network to be found.
     * @return Network info or {@code null} if the network was not found.
     */
    @Nullable
    public static NetworkInfo findNetwork(
            @NonNull List<NetworkInfo> networks, @Nullable String chainId, int coin) {
        if (networks.isEmpty() || TextUtils.isEmpty(chainId)) {
            return null;
        }
        return JavaUtils.find(
                networks,
                networkInfo -> networkInfo.chainId.equals(chainId) && networkInfo.coin == coin);
    }

    public static boolean isTestNetwork(String chainId) {
        return WalletConstants.KNOWN_TEST_CHAIN_IDS.contains(chainId);
    }
}
